src/overlays/actors/ovl_Bg_Last_Bwall/z_bg_last_bwall.c

```diff
@@ -28,57 +37,210 @@ const ActorInit Bg_Last_Bwall_InitVars = {
+    Vec3f spB4;
+    s32 i;
+
+    Matrix_RotateYS(this->dyna.actor.shape.rot.y, MTXMODE_NEW);
+    Matrix_RotateXS(this->dyna.actor.shape.rot.y, MTXMODE_APPLY);
+    Matrix_RotateZS(this->dyna.actor.shape.rot.y, MTXMODE_APPLY);
+    for (i = 0; i < 30; i++) {
+        switch (this->unk238) {
+            case 0:
+                spD0.x = 280.0f;
+                spD0.y = Rand_Centered() * 160.0f + 320.0f;
+                spD0.z = Rand_Centered() * 160.0f;
+                Matrix_MultVec3f(&spD0, &position);
+                Math_Vec3f_Sum(&this->dyna.actor.world.pos, &position, &position);
+                break;
+            case 1:
+                spD0.x = Rand_Centered() * 160.0f;
+                spD0.y = 400.0f;
+                spD0.z = Rand_Centered() * 160.0f;
+                Matrix_MultVec3f(&spD0, &position);
+                Math_Vec3f_Sum(&this->dyna.actor.world.pos, &position, &position);
+                break;
+        }
+        func_800BBFB0(play, &position, 50.0f, 2, Rand_ZeroOne() * 120.0f + 20.0f, Rand_ZeroOne() * 240.0f + 20.0f, 0);
+        spB4.x = Rand_ZeroOne() * 2.5f;
+        spB4.y = Rand_ZeroOne() * 2.5f + 1.0f;
+        spB4.z = Rand_ZeroOne() * 2.5f;
+        randVar = Rand_ZeroOne();
+        if (randVar < 0.2f) {
+            var_v0 = 0x60;
+        } else if (randVar < 0.6f) {
+            var_v0 = 0x40;
+        } else {
+            var_v0 = 0x20;
+        }
+        EffectSsKakera_Spawn(play, &position, &spB4, &position, -0x104, var_v0, 0x14, 0, 0, 0xA, 0, 0, 0x32, -1, 0x234,
+                             object_last_obj_DL_000098);
+    }
+}
+
+void func_80C187E4(BgLastBwall* this) {
+    this->actionFunc = func_80C187F8;
+}
+
+void func_80C187F8(BgLastBwall* this, PlayState* play) {
+    if (this->colliderTris.base.acFlags & 2) {
+        this->colliderTris.base.acFlags &= 0xFFFD;
+        Flags_SetSwitch(play, BGLASTBWALL_GET_SWITCHFLAGS(&this->dyna.actor));
+        func_80C1886C(this, play);
+    } else {
+        CollisionCheck_SetAC(play, &play->colChkCtx, &this->colliderTris.base);
+    }
+}
+
+void func_80C1886C(BgLastBwall* this, PlayState* play) {
+    this->actionFunc = func_80C18884;
+}
+
+void func_80C18884(BgLastBwall* this, PlayState* play) {
+    if (SubS_StartActorCutscene(&this->dyna.actor, this->nextCutscene, -1, 0)) {
+        func_80C188C4(this, play);
+    }
+}
+
+void func_80C188C4(BgLastBwall* this, PlayState* play) {
+    func_800C62BC(play, &play->colCtx.dyna, this->dyna.bgId);
+    this->dyna.actor.draw = NULL;
+    func_80C184EC(this, play);
+    Actor_PlaySfxAtPos(&this->dyna.actor, NA_SE_EV_WALL_BROKEN);
+    this->actionFunc = BgLastBwall_DoNothing;
+}
+
+void BgLastBwall_DoNothing(BgLastBwall* this, PlayState* play) {
+}
+
+void BgLastBwall_Update(Actor* thisx, PlayState* play) {
+    BgLastBwall* this = THIS;
+    this->actionFunc(this, play);
+}
+
+void BgLastBwall_Draw(Actor* thisx, PlayState* play) {
+    Gfx_DrawDListOpa(play, D_80C18A48[THIS->unk238].dList);
```

Comment:
Interestingly, the usual upcasting into its own variable doesn't match. Is there another way to do this?

Reply:
Can you create a scratch?

Reply:
https://decomp.me/scratch/GMR8S

Reply:
Interesting, it needs both: https://decomp.me/scratch/LAxrZ